src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,531 @@
+                // No disputes available, can't proceed
+                break;
+            }
+        }
+
+        console.log(`Collected ${disputeWindows.length} dispute windows`);
+
+        // Return payload with all available data
+        return {
+            latestForkGenesisSnapshot,
+            ...(disputeWindows.length > 0 && { disputeWindows }),
+            ...(latestStateProof && { stateProof: latestStateProof }),
+            ...(encodedState && { encodedState })
+        };
+    }
+
+    /**
+     * Fetch latest on-chain snapshot
+     */
+    private async fetchOnChainSnapshot(
+        channelId: ChannelId
+    ): Promise<StateSnapshot> {
+        // Fetch the latest on-chain snapshot from RPC node
+        // Assume it's true since it's on-chain
+        const currentOnChainSnapshot = StateSnapshot.from(
+            await this.mainRpcService.p2pManager.stateManager.stateChannelManagerContract.getStateSnapshot(
+                channelId
+            )
+        );
+        return currentOnChainSnapshot;
+    }
+
+    /**
+     * Run the payload in locally to verify it
+     */
+    private async verifyPayload(
+        channelId: ChannelId,
+        onChainSnapshot: StateSnapshot,
+        snapshotPayload: SnapshotPayload
+    ): Promise<{ isValid: boolean; error?: string; provenState?: any }> {
+        try {
+            const stateManager = this.mainRpcService.p2pManager.stateManager;
+
+            // Start with on-chain snapshot and traverse dispute windows until we reach the latest fork
+            let currentForkId = onChainSnapshot.forkId;
+            let isDisputed =
+                await stateManager.stateChannelManagerContract.isForkDisputed(
+                    channelId,
+                    currentForkId
+                );
+
+            while (isDisputed) {
+                // Check if reduced result already exists on-chain
+                const existingReducedResult =
+                    await stateManager.stateChannelManagerContract.getReducedResult(
+                        channelId,
+                        currentForkId
+                    );
+                if (existingReducedResult[0]) {
+                    // Traverse to the reduced fork
+                    currentForkId = existingReducedResult[0];
+                    isDisputed =
+                        await stateManager.stateChannelManagerContract.isForkDisputed(
+                            channelId,
+                            currentForkId
+                        );
+                    continue;
+                }
+
+                // No existing reduced result, use the disputes provided by participant
+                if (
+                    snapshotPayload.disputeWindows &&
+                    snapshotPayload.disputeWindows.length > 0
+                ) {
+                    // Pop the next dispute window
+                    const currentWindowDisputes =
+                        snapshotPayload.disputeWindows.shift()!;
```

Comment:
yep doesn't account for race condtion / the posibility that this disputeWindow could be reduced on-chain and skiped above